pageserver/src/tenant_mgr.rs

```diff
@@ -122,18 +122,20 @@ pub fn attach_local_tenants(
         );
         debug!("Timelines to attach: {local_timelines:?}");
 
-        let tenant = load_local_tenant(conf, tenant_id, remote_index);
-        {
-            match tenants_state::write_tenants().entry(tenant_id) {
-                hash_map::Entry::Occupied(_) => {
-                    error!("Cannot attach tenant {tenant_id}: there's already an entry in the tenant state");
-                    continue;
-                }
-                hash_map::Entry::Vacant(v) => {
-                    v.insert(Arc::clone(&tenant));
-                }
+        let mut tenants_accessor = tenants_state::write_tenants();
+        let tenant = match tenants_accessor.entry(tenant_id) {
+            hash_map::Entry::Occupied(o) => {
+                info!("Tenant {tenant_id} was found in pageserver's memory");
+                Arc::clone(o.get())
             }
-        }
+            hash_map::Entry::Vacant(v) => {
+                info!("Tenant {tenant_id} was not found in pageserver's memory, loading it");
+                let tenant = load_local_tenant(conf, tenant_id, remote_index);
```

Comment:
Was it intentional to switch the order of write lock to cover the local file reading? Cannot figure this out from the original issue nor the surrounding comments, maybe I missed some.

Looking at the `load_local_tenant` and assuming the file read could be done without the lock, so it would seem that only modifying the old lines 128..=130 would suffice...

Reply:
I'm not sure if it's formulated well enough in the codebase, but general model is to consider that if tenant/timeline is in memory, it has a correct set of files loaded to disk.

I've changed the code logic to do "if in memory, take that; load the whole tenant from scratch otherwise", so `load_local_tenant` has to be done only in the "not in memory" branch.
I guess I could have to do the loading outside of the lock, but considered that tenant has so few files to load from disk (a tenant config only), so this is ok.

Reply:
Ack, reading the single file is most likely ok while holding the lock. Resolving..